### ISSUE TITLE
[Tracking] Hot Module Replacement with webpack

### ISSUE THREAD

**Issue by sis0k0** [NativeScript/NativeScript#6398]
## Overview
Hot Module Replacement (HMR) is a feature provided by webpack. It makes it possible to replace modules in a bundled application while the application is running.

Integrating HMR within NativeScript will improve the development experience by:
- reducing the time to apply the updates when changing files;
- preserving the application state between reloads.

## Design
Implementation details and motivation of the feature can be found in its [design document](https://docs.google.com/document/d/1Fu6w_wjr9NwinFs2PWU7e2L3z5TbGSxNYh0WNXVSPg0/edit?usp=sharing).

### NativeScript 5.0
- [X] Demo applications.
    - [X] JavaScript https://github.com/NativeScript/js-ns-hmr
    - [X] Angular https://github.com/NativeScript/ng-ns-hmr 
- [x] **{N}-webpack** Integrate the `HotModuleReplacement` plugin with `nativescript-dev-webpack`.
- [x] **{N}-webpack** Template for HMR runtime and logic that triggers the hot update on change.
- [X] **{N}-webpack** Loaders for invoking `reloadPage` on changes in JS/TS-only applications.
- [X] **{N}-webpack** Add Vue template. https://github.com/NativeScript/nativescript-dev-webpack/pull/676
- [x] **{N}-CLI** Flag for enabling HMR.
- [x] **{N}-CLI** Sync the backup files and restart the application when the hot update fails.
- [X] **{N}-CLI** Integrate with `tns preview`.
- [X] Integrate with NativeScript Playground.
- [X] **{N}-Angular** Modify the bootstrap API to allow for dynamic module refetching.
- [X] **{N}-Angular** Application logic for bootstrapping on changes.
- [ ] **{N}-Angular** Research the Angular compiler API for ways to improve the application reload logic in NativeScript Angular application.
- [ ] **{N}-webpack** Loaders for dynamically inserting the reload logic in NativeScript Angular applications.
- [ ] Provide a guide on persisting the application state between reloads.
- [ ] Documentation.
	- [ ] Update the {N}-webpack article.
	- [ ] Mention the `hmr` flag in the {N}-CLI section.

### After NativeScript 5.0
The possible improvements are explained in the [design doc](https://docs.google.com/document/d/1Fu6w_wjr9NwinFs2PWU7e2L3z5TbGSxNYh0WNXVSPg0/edit?usp=sharing).

- [ ] **{N}-CLI** Research ways to sync updates without FS
- [ ] **{N}-modules** Apply new styles at runtime
- [ ] **{N}-modules** Apply new UI at runtime
- [ ] **{N}-modules** Improve the reload strategy when the root view is not `Page`, but `TabView`, `SideDrawer`, etc.
- [ ] **{N}-modules** Come up with a reload strategy for custom components.
- [ ] **{N}-Angular** Research ways to persist route between reloads.

## Contributing
HMR provides us with so many possibilities!
If you want to work on a task from the list, tell us and we will do our best to help you!
If you have another great idea on how to make HMR even cooler, share it in the comments!

## Discussion
Please, don't report problems here. Instead, open a new issue and link it to this one.
Let's use this discussion for suggestions and improvement ideas. We would love to here from you!

**Comment by shiv19** [comment]
@tsonevn Navigation state is also not maintained after HMR.

do you want me to create a new issue to track it?

**Comment by sis0k0** [comment]
@shiv19, 10x for the feedback! You don't need to log an issue. I will add `Research ways to persist nav history` to the list of possible improvements.

TLDR;
This is the expected behavior of the [reloadPage](https://github.com/NativeScript/NativeScript/blob/c8c0be76846446124ce4cb27f59dda99d4898731/tns-core-modules/ui/frame/frame.android.ts#L77) function that's used for applying the changes.

**Comment by sis0k0** [comment]
Thanks for the feedback, @shiv19! I will add the item to the list of future improvements. You don't need to log it in a separate issue.

TLDR;
This is the expected behavior of the [reloadPage](https://github.com/NativeScript/NativeScript/blob/c8c0be76846446124ce4cb27f59dda99d4898731/tns-core-modules/ui/frame/frame.android.ts#L77) function that we're using for applying the changes.

**Comment by larssn** [comment]
Is HMR _supposed_ to tear down the entire angular app when reloading? Thought it was supposed to maintain the state? We're seeing ngOnDestroy getting called for everything on reloads, and it's causing problems.

Can anyone clarify?

**Comment by MarkPieszak** [comment]
Typically when trying to do HMR like this, you might need to check for `module['hot']` and dispose the previous one.

(As an example I do for aspnetcore & angular integration with HMR)
```ts
// // Enable either Hot Module Reloading or production mode
if (module['hot']) {
  module['hot'].accept();
  module['hot'].dispose(() => {
    modulePromise.then(appModule => appModule.destroy());
  });
} else {
  enableProdMode();
}
```

**Comment by vchimev** [comment]
Updated for 5.2 release.

**Comment by hettiger** [comment]
> [  ] {N}-modules Improve the reload strategy when the root view is not Page, but TabView, SideDrawer, etc.

Are there any workarounds? When can we expect this to work out of the box?

**Comment by vchimev** [comment]
Hey @hettiger,

This has been implemented for NativeScript applications with JavaScript and TypeScript:
- related issue: https://github.com/NativeScript/NativeScript/issues/7145
- available with `tns-core-modules@next`
- official release with NativeScript 5.4.0

**Comment by vchimev** [comment]
Tracking issue is updated for 5.4 release.

**Comment by shiv19** [comment]
I tried it today and it works like a dream! Love it <3

**Comment by vahidvdn** [comment]
Hey guys. When I close the app and open it again, every changes with `--hmr` is gone. But this doesn't happen without `--hmr`. Any idea?

**Comment by vchimev** [comment]
Hey @vahidvdn,

There was a regression in some of the *next* versions of `nativescript-dev-webpack` and that could be the reason. Could you please make sure you use the most recent `latest` or `next` version?

Let me know if still experience this problem.

**Comment by lambourn** [comment]
I understand that {N} + Angular + HMR means that the application will bootstrap on change. 

Does this imply that objects created by Angular's DI / bootstrap mechanisms will get recreated from scratch? Or will a @Injectable object be "the same" before and after HMR executes?

Right now I see lots of (non-UI) objects getting re-created after HMR.

**Comment by vahidvdn** [comment]
Hi @vchimev 
This still happens in nativescript 6.1. I just update my cli and created a project with Angular tab template.
Now `hmr` is enabled by default. I save some changes, close the app, then I open it, I don't see the changes. And due to closing the app, I can't save new changes anymore.

**Comment by Jonatthu** [comment]
@vahidvdn mine used to do the same until I removed the come on maint.ns.ts that used to listen for hmr module flag.

**Comment by vahidvdn** [comment]
@Jonatthu Thanks, but I'm afraid I didn't catch on you. You just turned off the hrm?

**Comment by Jonatthu** [comment]
@vahidvdn I was having problems with my hmr until I commented out this piece of code.
After I removed this legacy code, my start working with 6.1

**Comment by trajano** [comment]
Would using mettro instead of webpack make things easlier since they seem to have solved the situation with Expo

**Comment by shirakaba** [comment]
> Would using mettro instead of webpack make things easlier since they seem to have solved the situation with Expo

@trajano It's not realistic to port our very webpack-reliant ecosystem over to Metro, which in any case has far worse documentation and is a far less open ecosystem. Callstack are in fact building solutions to [move away from Metro to Webpack](https://github.com/callstack/repack), which I believe Expo are interested in (as it would allow a common and well-understood bundler for both React Native and React Native Web).

The bundler itself is a relatively minor factor in how good the HMR experience is.

**Comment by trajano** [comment]
Thanks @shirakaba  I am just trying to see how to get all of this working to the same level as we have with Expo and React-Native where we can make changes and it updates the app automatically.  I see announcements for HMR with Nativescript but I could never get it working and a few StackOverflow questions lead to issues that lead here.

**Comment by shirakaba** [comment]
@trajano what specifically is not working, and which flavour of NativeScript are you using (Angular, Vue, etc.), and what version of Webpack and what version of NativeScript?

**Comment by trajano** [comment]
@shirakaba I"ll start a new issue so I don't make this go too long.
